[Flaky] k8s.io/apiserver/pkg/server/filters - TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter

**Which jobs are flaking**:

* ci-kubernetes-cached-make-test
* ci-kubernetes-generate-make-test-count10
* pull-kubernetes-unit

**Which test(s) are flaking**:

* TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter

**Testgrid link**:

* https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-generate-make-test-count10&include-filter-by-regex=TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter&width=20
* https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-cached-make-test&include-filter-by-regex=TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter&width=20
* https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit&include-filter-by-regex=TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter&width=20

**Reason for failure**:

Data race:

```
==================
WARNING: DATA RACE
Read at 0x00c0006c83d0 by goroutine 199:
  k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset.(*queueSet).StartRequest.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:300 +0x12d

Previous write at 0x00c0006c83d0 by goroutine 183:
  k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset.(*queueSet).setConfiguration()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:207 +0x10b
  k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset.(*queueSetCompleter).Complete()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:160 +0x117
  k8s.io/apiserver/pkg/util/flowcontrol.(*cfgMeal).finishQueueSetReconfigsLocked()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:650 +0x72a
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).lockAndDigestConfigObjects()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:490 +0x1b6
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).digestConfigObjects()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:397 +0xa4
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).syncOne()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:355 +0x5af
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).processNextWorkItem.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:325 +0x14f
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).processNextWorkItem()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:335 +0x84
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).runWorker()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:311 +0x56
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).runWorker-fm()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:310 +0x2b
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xba
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x114
  k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x5a

Goroutine 199 (running) created at:
  k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset.(*queueSet).StartRequest()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:290 +0x314
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).startRequest()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:807 +0xffe
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Handle()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:146 +0xa4
  k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:130 +0x8cd
  net/http.HandlerFunc.ServeHTTP()
      /Users/liggitt/.gvm/gos/go1.16/src/net/http/server.go:2069 +0x51
  k8s.io/apiserver/pkg/server/filters.newHandlerChain.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:849 +0x3f6
  net/http.HandlerFunc.ServeHTTP()
      /Users/liggitt/.gvm/gos/go1.16/src/net/http/server.go:2069 +0x51
  k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:107 +0x10c

Goroutine 183 (running) created at:
  k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Run()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:300 +0x425
  k8s.io/apiserver/pkg/server/filters.startAPFController.func1()
      /Users/liggitt/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:761 +0x4c
==================
I0325 20:34:13.498081   11980 apf_controller.go:303] Shutting down API Priority and Fairness config worker
I0325 20:34:13.513730   11980 apf_controller.go:303] Shutting down API Priority and Fairness config worker
--- FAIL: TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter (0.23s)
    --- FAIL: TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter/priority_level_concurrency_is_set_to_1,_queue_length_is_1,_first_request_should_time_out_and_second_(enqueued)_request_should_time_out_as_well (5.16s)
        priority-and-fairness_test.go:693: Waiting for the request: "/request/first/time-out-as-designed" to time out
        priority-and-fairness_test.go:712: Waiting for the inner handler of the request: "/request/first/time-out-as-designed" to complete
        priority-and-fairness_test.go:732: Waiting for the controller to shutdown
        testing.go:1093: race detected during execution of test
    --- FAIL: TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter/priority_level_concurrency_is_set_to_1,_inner_handler_panics_after_the_request_times_out (5.17s)
        priority-and-fairness_test.go:541: Waiting for the request: "/request/time-out-as-designed" to time out
        priority-and-fairness_test.go:548: Waiting for the inner handler of the request: "/request/time-out-as-designed" to complete
        priority-and-fairness_test.go:559: Waiting for the controller to shutdown
        testing.go:1093: race detected during execution of test
FAIL
```

Reproduced in ~2% of runs with [stress](https://pkg.go.dev/golang.org/x/tools/cmd/stress):

```
go test k8s.io/apiserver/pkg/server/filters -race -c
stress ./filters.test -test.run TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter/
...
193 runs so far, 5 failures
204 runs so far, 5 failures
216 runs so far, 5 failures
...
```

Triage board:

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter

/sig api-machinery

--- a/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
@@ -285,6 +285,11 @@ func (qs *queueSet) StartRequest(ctx context.Context, hashValue uint64, flowDist
 	// request's context's Done channel gets closed by the time
 	// the request is done being processed.
 	doneCh := ctx.Done()
+
+	// Retrieve the queueset configuration name while we have the lock
+	// and use it in the goroutine below.
+	configName := qs.qCfg.Name
+
 	if doneCh != nil {
 		qs.preCreateOrUnblockGoroutine()
 		go func() {
@@ -297,7 +302,7 @@ func (qs *queueSet) StartRequest(ctx context.Context, hashValue uint64, flowDist
 			// known that the count does not need to be accurate.
 			// BTW, the count only needs to be accurate in a test that
 			// uses FakeEventClock::Run().
-			klog.V(6).Infof("QS(%s): Context of request %q %#+v %#+v is Done", qs.qCfg.Name, fsName, descr1, descr2)
+			klog.V(6).Infof("QS(%s): Context of request %q %#+v %#+v is Done", configName, fsName, descr1, descr2)
 			qs.cancelWait(req)
 			qs.goroutineDoneOrBlocked()
 		}()

